### ISSUE TITLE
function plotting code

### ISSUE THREAD

**Issue by RobinSchmidt** [RobinSchmidt/RS-MET#298]
i was recently experimenting a bit with creating contour plots in GNUPlotCPP and i was a bit underwhelmed by the results so i sat down and implemented my own, low-level contour plotting algorithm. the cool thing is that it can be applied to any image - the pixel brightness values are interpreted as height-values and the algorithm draws the lines for contour levels and - in another variation - fill the areas between the contour lines with selectable colors. for demo-purposes, i created some 2D function (involving sinusoidal and polynomial components in x and y) and plotted its contour lines and fills.

![image](https://user-images.githubusercontent.com/1833099/74971310-2f736600-5420-11ea-9eb0-ab0903227b4b.png)

![image](https://user-images.githubusercontent.com/1833099/74971335-3ac69180-5420-11ea-9e56-d85e3d5167a3.png)

![image](https://user-images.githubusercontent.com/1833099/74971355-43b76300-5420-11ea-9efb-7069d89d10ad.png)

for the filling algorithm, the number of colors is independent from the number of levels - the colors just wrap around - her, the same function with 3 colors/gray-levels:

![image](https://user-images.githubusercontent.com/1833099/74971867-10290880-5421-11ea-8f4f-e97eee976c86.png)

**Comment by RobinSchmidt** [comment]
the most tricky part was (as always) the anti-aliasing - for comparison, here are the lines and the 3-color fills without anti-aliasing:
![image](https://user-images.githubusercontent.com/1833099/74972106-76159000-5421-11ea-9169-cfa05484edf0.png)

![image](https://user-images.githubusercontent.com/1833099/74972142-82015200-5421-11ea-8477-1e8176d83f0a.png)

**Comment by RobinSchmidt** [comment]
my main goal was to find a nice way to visualize complex functions, i.e. functions that take a complex number as input and produce a complex number as output. my idea is to use a contour plot for the real and imaginary component of the output and use one for the red channel and the other for the blue or green channel. here is f(z) = z^2 with real -> red, imaginary -> blue using contour filling:
![image](https://user-images.githubusercontent.com/1833099/74972638-60549a80-5422-11ea-9a47-50ae116c2df9.png)
areas of high red have high real part, areas with high blue have high imaginary part, in magenta areas, both part are high, in black areas, both are low. i've not yet seen complex functions being plotted/visualized this way.  f(z) = z^2 has real part: x^2 - y^2 and imaginary part 2*x*y. both parts have hyperbolas as level lines - but rotated by 45° with respect to one another

**Comment by elanhickler** [comment]
So the use case is to help visualize complex numbers? Is it for debugging algorithms?

**Comment by RobinSchmidt** [comment]
f(z) = z^3 looks like this:
![image](https://user-images.githubusercontent.com/1833099/74973139-33ed4e00-5423-11ea-9ac4-083c6f97a766.png)

there were a couple of spin-offs. i can now also plot  implicit curves

![image](https://user-images.githubusercontent.com/1833099/74973208-52ebe000-5423-11ea-8a6b-d3bd1c0f81a6.png)

...parametric curves and soon also simply functions. ...one nice thing is that to create plots for debugging and/or demonstartion purposes, i'll soon won't necessarily depend on GNUPlot anymore. there may be other benefits as well. ...i could perhaps come up with a sort of "PrettyPlot" application...

**Comment by elanhickler** [comment]
that does look pretty.

How about visualizing chaos?

**Comment by RobinSchmidt** [comment]
> So the use case is to help visualize complex numbers? Is it for debugging algorithms?

yes, this was my initial goal. mostly, because i want to understand them better and get a better intuitive feeling for them. complex functions occur in signal processing all over the place - mostly in form of rational transfer functions of filters. there are various ways to visualize them (for example, 3blue1brown likes to draw a grid in the input plane and show the resulting grid of curves in the output plane), but i've not yet seen it being done like this. 

> How about visualizing chaos?

hmm...how? i guess, i could draw mandelbrot-sets and bifurcation diagrams and the likes - is that what you mean?

**Comment by RobinSchmidt** [comment]
i've also improved the anti-aliasing in my line drawing code - here without anti-aliasing, clearly featuring jaggies:
![image](https://user-images.githubusercontent.com/1833099/74975553-86306e00-5427-11ea-9a25-d69abe26da99.png)

here the old anti-aliasing - the jaggies are gone - but there's a new artifact that i call "twisties":
![image](https://user-images.githubusercontent.com/1833099/74975646-ae1fd180-5427-11ea-803f-ae1bdbabfe3d.png)

and here the new code:
![image](https://user-images.githubusercontent.com/1833099/74975744-d8718f00-5427-11ea-8ae8-f2d331f022ed.png)

**Comment by elanhickler** [comment]
>>How about visualizing chaos?
>hmm...how? i guess, i could draw mandelbrot-sets and bifurcation diagrams and the likes - is that what you mean?

I dunno what I mean.

May be unrelated: Can you improve my chaotic algorithms which I use to create analog-esque filters? Can you solve explosions? Essentially parameters need to be limited based on samplerate.
